src/plugins/internalnotification/macros/addType.h

```diff
@@ -0,0 +1,71 @@
+#ifndef TYPE
+#error "You have to #define TYPE, TYPE_NAME and TO_VALUE before including the addType supermacro"
+#endif
+#ifndef VALUE_TYPE
+// use type as default if not set
+#define VALUE_TYPE TYPE
+#endif
+#ifndef TYPE_NAME
+#error "You have to #define TYPE, TYPE_NAME and TO_VALUE before including the addType supermacro"
+#endif
+#ifndef TO_VALUE
+#error "You have to #define TYPE, TYPE_NAME and TO_VALUE before including the addType supermacro"
+#endif
+#ifndef CHECK_CONVERSION
+#define CHECK_CONVERSION 1
+#endif
+
+#define CONCAT(X, Y) CONCAT2 (X, Y)
```

Comment:
Is this needed as macro?

Reply:
Yes. It concatenates two tokens. It was not necessary before since it was already encapsulated in a macro before expansion. I'm not sure how to describe, so here is an example :)

```C
// in supermacro.h
#define CONCAT(X, Y) CONCAT2 (X, Y)
#define CONCAT2(X, Y) X##Y
#define FUNC_NAME(TYPE_NAME) CONCAT (internalnotificationRegister, TYPE_NAME)
#define FUNC_NAME2(TYPE_NAME) internalnotificationRegister##TYPE_NAME
#define NESTED(TYPE_NAME) FUNC_NAME2 (TYPE_NAME)

FUNC_NAME (TYPE_NAME)
FUNC_NAME2 (TYPE_NAME)
NESTED (TYPE_NAME)

// in some.c
#define TYPE_NAME Int
#include "macros/supermacro.h"
// => gcc -E some.c
internalnotificationRegisterInt
internalnotificationRegisterTYPE_NAME
internalnotificationRegisterInt
```

Before the "supermacro", the `NESTED` macro was the macro that generated functions. A solution would be to add another macro (e.g. `#define FUNC_SIGNATURE(TYPE_NAME) void FUNC_NAME(TYPE_NAME) (void)` but the `CONCAT` macro is more versatile.

Reply:
Okay, then we should also add `ELEKTRA_CONCAT` next to `ELEKTRA_STRINGIFY`.